pkg/controller/direct/apikeys/apikeyskey/apikeyskey_controller.go

```diff
@@ -0,0 +1,296 @@
+	Ignore("updateTime"),
+	Ignore("deleteTime"),
+	Ignore("etag"),
+	Ignore("name"),        // TODO: Should be ResourceID?
+	Ignore("annotations"), // TODO: Should not ignore
+).
+	MapNested(&pb.Restrictions{}, &v1alpha1.KeyRestrictions{}, "apiTargets",
+		"androidKeyRestrictions", "browserKeyRestrictions", "iosKeyRestrictions", "serverKeyRestrictions").
+	MapNested(&pb.AndroidKeyRestrictions{}, &v1alpha1.KeyAndroidKeyRestrictions{}, "allowedApplications").
+	MapNested(&pb.AndroidApplication{}, &v1alpha1.KeyAllowedApplications{}, "packageName", "sha1Fingerprint").
+	MapNested(&pb.ApiTarget{}, &v1alpha1.KeyApiTargets{}, "methods", "service").
+	MapNested(&pb.BrowserKeyRestrictions{}, &v1alpha1.KeyBrowserKeyRestrictions{}, "allowedReferrers").
+	MapNested(&pb.IosKeyRestrictions{}, &v1alpha1.KeyIosKeyRestrictions{}, "allowedBundleIds").
+	MapNested(&pb.ServerKeyRestrictions{}, &v1alpha1.KeyServerKeyRestrictions{}, "allowedIps").
+	MustBuild()
+
+type adapter struct {
+	projectID string
+	location  string
+	keyID     string
+
+	desired *v1alpha1.APIKeysKey
+	actual  *v1alpha1.APIKeysKey
+
+	gcp *apikeys.Client
+}
+
+func (m *model) client(ctx context.Context) (*apikeys.Client, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	// if m.config.Endpoint != "" {
+	// 	opts = append(opts, option.WithEndpoint(m.config.Endpoint))
+	// }
+
+	gcpClient, err := apikeys.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building apikeys client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	gcp, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &v1alpha1.APIKeysKey{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	projectID := obj.Spec.ProjectRef.External
+	if projectID == "" {
+		return nil, fmt.Errorf("unable to determine project")
+	}
+
+	// TODO: Use name or request resourceID to be set on create?
+	keyID := ValueOf(obj.Spec.ResourceID)
+	if keyID == "" {
+		return nil, fmt.Errorf("unable to determine resourceID")
+	}
+
+	location := "global"
+
+	return &adapter{
+		projectID: projectID,
+		location:  location,
+		keyID:     keyID,
+		desired:   obj,
+		gcp:       gcp,
+	}, nil
+}
+
+func ValueOf[T any](p *T) T {
+	var v T
+	if p != nil {
+		v = *p
+	}
+	return v
+}
+
+func (a *adapter) Find(ctx context.Context) (bool, error) {
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
```

Comment:
Would it make more sense to add things like keyID != "" check into the fullyQualifiedName() method and add an error return?